### ISSUE TITLE
java.lang.NoSuchMethodError when classes are in different modules

### ISSUE THREAD

**Issue by lallam** [Kotlin/kotlinx.serialization#576]
Runtime Crash

java.lang.NoSuchMethodError: No direct method <init>(ILkotlinx/serialization/SerializationConstructorMarker;)V in class Lcom/xxxx/common/core/Model; or its super classes (declaration of 'com.xxxx.common.core.Model' appears in /data/app/com.xxxx.demo-PV-n86-hzEl-eyc8UqbACQ==/base.apk!classes7.dex)
10-11 15:30:10.048 E/AndroidRuntime(25760): at com.xxxx.reg.data.User.<init>(Unknown Source:9)
10-11 15:30:10.048 E/AndroidRuntime(25760): at com.xxxx.reg.data.User$$serializer.deserialize(Unknown Source:624)
10-11 15:30:10.048 E/AndroidRuntime(25760): at com.xxxx.reg.data.User$$serializer.deserialize(User.kt:17)
10-11 15:30:10.048 E/AndroidRuntime(25760): at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:33)

**To Reproduce**

@Serializable
abstract class Model(
    var uniqueId: String = "",
}

@Serializable
data class User(
        var userName: String = "",
        var dateOfBirth: String = "",
) : Model() 

**Library A** - Model
**Library B** - User

I am getting the above crash.

BUT if I move the Model from Library A to Library B NO issues.
[Just an FYI]
if the variable uniqueId from Model is removed no issue
if @Serializable is removed from Model no issues 

**Expected behavior**
Not crash or throw fatal error.

**Environment**
 - Kotlin version: [e.g. 1.3.50]
 - Library version: [e.g. 0.13.0]

Updated Post with More comments and Link to Code - https://github.com/Kotlin/kotlinx.serialization/issues/576#issuecomment-637581639

**Comment by lallam** [comment]
any update on this or verfied this as a known issue or has been fixed?

a class extended from either same library or a different library should yield same results.

**Comment by sandwwraith** [comment]
Sorry, seems like this issue slipped out of my sight. Can confirm this bug on 1.3.60 with mutli-module (library/app) setup

**Comment by lallam** [comment]
Yes we are using 0.14.0

this is still an issue.

**Comment by JavierSegoviaCordoba** [comment]
Kotlin 1.3.61 and Kotlin Serialization 0.14.0 is working for me, but with Kotlin 1.3.70:

```
Caused by: java.lang.NoSuchMethodError: No direct method <init>(Ljava/lang/String;Lkotlinx/serialization/internal/GeneratedSerializer;I)V in class Lkotlinx/serialization/internal/SerialClassDescImpl; or its super classes (declaration of 'kotlinx.serialization.internal.SerialClassDescImpl' appears in /data/app/com.javiersc.gitGut-9pedF-cZ8kBfR8wDIz6hJg==/base.apk)
        at com.javiersc.gitGut.models.UserDTO$$serializer.<clinit>(UserDTO.kt:7)
        at com.javiersc.gitGut.models.UserDTO$Companion.serializer(Unknown Source:0)
```

**Comment by b95505017** [comment]
I can confirm `kotlin 1.3.70` is not compatible with `serialization 0.14.0` as well.

**Comment by JavierSegoviaCordoba** [comment]
I tried the last eap 0.20.0 and it is working

**Comment by sandwwraith** [comment]
Yes, unfortunately, there's a breaking change in 1.3.70 plugin. We'll release compatible version of serialization in the following next few hours. Meanwhile, you can try eap builds from bintray: https://bintray.com/kotlin/kotlinx/kotlinx.serialization.runtime (e.g. `0.20.0-1.3.70-eap-274-2`)

**Comment by brianguertin** [comment]
Hmm, even the latest 0.20.0 is still failing for me with "no direct method <init>" error

**Comment by sandwwraith** [comment]
Original issue, which is a bug, not an incompatibility with Kotlin, is still present. @JavierSegoviaCordoba's issue was #736

**Comment by JavierSegoviaCordoba** [comment]
@sandwwraith maybe I haven't the exact issue because the eap version fixed it for me.

**Comment by lallam** [comment]
Updated the versions to

Kotlin - 1.3.70
Kotlin Serialization - 0.20.0

Confirming that Issue is still there.

**Comment by enigmaCoder** [comment]
@sandwwraith This issue is not the same as @JavierSegoviaCordoba is mentioning fixed.
This issue is rather an issue in the Serialization annotation processor which doesn't allow inheritance of the Serialization property when working in an multi module project and serialization property are failing to be serialized at runtime. As you can see it throws an java.lang.NoSuchMethod error as it cannot process 
However while working in an non multi module project this issue of inheritance is not there. If it was an Kotlin issue it would have given in an non multi module project as well if I am not wrong

**Comment by enigmaCoder** [comment]
@sandwwraith Also I am not sure but I believe the fix would need @Inherit added in the @Serialzable annotation so that this can be inherited across modules

**Comment by enigmaCoder** [comment]
@Krondo Issue still persists. Hence reopening this issue

**Comment by enigmaCoder** [comment]
@Krondo The issue mentioned in #763 is not the same.

**Comment by DevSrSouza** [comment]
I had the same issue

**Comment by bilalekremharmansa** [comment]
I might have a workaround..

kotlin: 1.3.70, kotlinx-serialization: 0.20.0

I was running my newly created project on IntelliJ ever since i got "java.lang.NoSuchMethodError". I figure out, kotlinx plugin is not get activated when I run my test classes in IDE. I need to run `mvn clean compile` and here we go;

[INFO] --- kotlin-maven-plugin:1.3.70:compile (compile) @ healthcheck ---
[INFO] Applied plugin: 'kotlinx-serialization'

so the serializer() function will be generated and exception will be gone right after you see the lines above. Hope it's solves your problem as well.

**Comment by MrPowerGamerBR** [comment]
This bug seems to happen only in JVM targets, I'm able to serialize/deserialize with no issues on my JS target but on my JVM target that bug happens.

**Comment by SanjayKushwah0601** [comment]
kotlin: 1.3.70, kotlinx-serialization: 0.20.0

Fixed my issue

**Comment by enigmaCoder** [comment]
I see how ppl are commenting that the issues are fixed in latest kotlin and kotlin serilzation plugin. Let me remind you guys to please understand the issue before commenting it's fixed and stop misleading everyone. 

And NO THIS ISSUE is not fixed in kotlin: 1.3.70, kotlinx-serialization: 0.20.0

**Comment by sandwwraith** [comment]
Changing the Kotlin version forces clean building, which fixes the issue sometimes (when it's caused by incremental compilation in one module).

**Comment by lallam** [comment]
Issue is NOT fixed.
Tested: 
Kotlin 1.3.72
kotlinx-serialization: 0.20.0

I have pushed test code in this repo

https://github.com/lallam/broke.git

project has the orginal code outlined in original post https://github.com/Kotlin/kotlinx.serialization/issues/576#issue-506005603

app crashes at RUNTIME on line 32 during deserialization in MainActivity [it is a very simple code]

to illustrate the problem.

in **libraryb**, i have also duplicated Model as **ModelX**, and inherited the User class
**User** - if you uncomment **line 8** and comment **line 9**. problem goes away.

This is a multi project/module and/or compiler setup issue

**Comment by mmr118** [comment]
I'm new to android. @lallam, can you (or anyone) ELI5 how to set **kotlinx-serialization**?

My build script is:
```
buildscript {
    ext.kotlin_version = '1.3.70'
    repositories {
        google()
        jcenter()
        maven { url "https://jitpack.io" }
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:4.0.0'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}
```

**Comment by dgomolka** [comment]
Kotlin 1.3.72
kotlinx-serialization: 0.20.0
java.lang.NoSuchMethodError: No direct method (ILkotlinx/serialization/SerializationConstructorMarker;)V

still occurs

**Comment by mschirmacher** [comment]
Is there any progress on this one? This issue drives the library completly useless...

**Comment by lallam** [comment]
Reference: https://github.com/Kotlin/kotlinx.serialization/issues/576#issue-506005603
https://github.com/Kotlin/kotlinx.serialization/issues/576#issuecomment-637581639

this has become a critical issue for us. we are currently using a VERY crude method to over come this temporarily.

Library A - Model

Library B - Model extends **LibraryA**.Model
                - User extends **LibraryB.Model**

Library C - Model extends **LibraryA**.Model
                - Device extends **LibraryC.Model**

Model classes in LibraryB and LibraryC are identical to LibraryA.Model and override the members. 
We are hoping that once the issue is fixed, we could DELETE these duplicate classes and extend User and Device classes directly from LibraryA.Model.

or other alternative is to abandon kotlin serialization altogether and refactor it to to GSON. fingers crossed.

**Comment by sandwwraith** [comment]
https://github.com/Kotlin/kotlinx.serialization/issues/378#issuecomment-663134060

**Comment by lallam** [comment]
thanks @sandwwraith we will test and confirm.

**Comment by IgnacioGarcia198** [comment]
Hello, this is happening to me with Kotlin 1.4 and org.jetbrains.kotlinx:kotlinx-serialization-core-jvm:1.0.0-RC which is included in Retrofit dependency (I could not use org.jetbrains.kotlinx:kotlinx-serialization-runtime-jvm:1.0-M1-1.4.0-rc since it creates dependency conflict).

**Comment by sandwwraith** [comment]
@IgnacioGarcia198 Do you have a sample project with reproducer?

**Comment by IgnacioGarcia198** [comment]
@sandwwraith I posted a project in my GitHub repository in order to reproduce it. Here is the link.
https://github.com/IgnacioGarcia198/BikeSharing-Android
If you try the test SerializationTest.kt, you will see it is working, but when you run the app the problem is with Retrofit. Please if you notice something I am doing wrong tell me ;)

**Comment by ronjunevaldoz** [comment]
Any update on this issue or example work around?

**Comment by pnemonic78** [comment]
maybe it has something to do with proguard obfuscation?

**Comment by enigmaCoder** [comment]
@IgnacioGarcia198 don't use the serializer factory of retrofit instead accept the Retrofit response as Response Body objects and the write your own serializing extension api and convert the response body received to the required module using the modules serializer using kotlin serialization.

**Comment by Arushi129** [comment]
Hi, I have the same error on my KMM project.

java.lang.NoSuchMethodError: No direct method (Ljava/lang/String;Lkotlinx/serialization/internal/GeneratedSerializer;I)V in class Lkotlinx/serialization/internal/SerialClassDescImpl; or its super classes (declaration of 'kotlinx.serialization.internal.SerialClassDescImpl' appears in /data/app/com.example.mykmmclient12april-eQ9TG4Mg_9sDGC-WYU26Eg==/base.apk)
at org.openapitools.client.models.Order$$serializer.(Order.kt:28)
at org.openapitools.client.models.Order$Companion.serializer(Unknown Source:0)
at org.openapitools.client.infrastructure.ApiClient$Companion.setMappers(ApiClient.kt:64)
at org.openapitools.client.infrastructure.ApiClient$Companion.access$setMappers(ApiClient.kt:57)
at org.openapitools.client.infrastructure.ApiClient$serializer$2.invoke(ApiClient.kt:41)
at org.openapitools.client.infrastructure.ApiClient$serializer$2.invoke(ApiClient.kt:28)
at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
at org.openapitools.client.infrastructure.ApiClient.getSerializer(Unknown Source:2)
at org.openapitools.client.infrastructure.ApiClient.jsonRequest(ApiClient.kt:148)
at org.openapitools.client.apis.StoreApi.placeOrder(StoreApi.kt:171)
at com.example.mykmmclient12april.MainActivity.callApi(MainActivity.kt:44)
at com.example.mykmmclient12april.MainActivity$onCreate$1.invokeSuspend(MainActivity.kt:29)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
at android.os.Handler.handleCallback(Handler.java:873)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:193)

Order Class -
@serializable
data class Order (
@SerialName(value = "id") val id: kotlin.Long? = null,
@SerialName(value = "petId") val petId: kotlin.Long? = null,
@SerialName(value = "quantity") val quantity: kotlin.Int? = null,
@SerialName(value = "shipDate") val shipDate: kotlin.String? = null,
/* Order Status */
@SerialName(value = "status") val status: Order.Status? = null,
@SerialName(value = "complete") val complete: kotlin.Boolean? = null
) {

Environment:

Kotlin version: [1.3.72]
serialization plugin version:[0.12.0]
IDE version [Android Studio 4.1.1]
gradle version:[6.5]

I doesn't want to update the kotlin and plugin version as it may affect my other projects. Please help.

**Comment by enigmaCoder** [comment]
Okay 
So first of, if you are using the @Serializable annotations you don't need to again declare @SerialName as kotlin serialization will automatically use the assigned one

Next is your "Order.Status" class serializable? Just check if it's not and make it serializable.

Next this problem would only happen if you are extending a serializable class which is in a seperate module to a Serializable class which is in a different module, this problem would arise.
This problem however is not there if both the classes exist in the same module.

So assuming you are facing the above problem, following are the solutions 
1.) You  must make sure both the serializable classes available in the same module make a copy of the class you are extending and keep a copy of it 
2. ) Instead of defining a common class provide a common interface and extend and implement it in all the classes needed
3.) Develop a reflection based serialization mechanism or use gson to serialize those modules which are having intermodule dependency

Hope this helps. Please share a sample piece of the above problem, so that we can take a closer look at the problem you are facing

**Comment by Arushi129** [comment]
Hi,

Thanks for the reply.
I am following the mentioned link to create KMM project using open api generator tool(provided yaml file only) - 1-)To create open api client - https://www.reddit.com/r/Kotlin/comments/d9xndy/openapi_generators_kotlin_client_generator_now/
2-) To download jar file - https://github.com/OpenAPITools/openapi-generator#13---download-jar

But the "kotlin version" in the generated open api client project from the above link is - "1.3.50" which is very old and not compatible with KMM latest version which is "1.4.32".

Note: I am trying to access the kotlin files or methods from swift ios standalone project.

**Comment by shanshin** [comment]
Not reproduced with Kotlin `2.0.20` and serialization runtime `1.7.3`
